### ISSUE TITLE
Should compilerError, compilerWarning, compilerAssert be in a module named by topic?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#16809]
For https://github.com/Cray/chapel-private/issues/1215

Today compilerError, compilerWarning, compilerAssert are in the Builtins module. Should these be moved to a module with a module name reflecting the topic? E.g. `CompilationDiagnostics` ? 

Note, this is separate from the question of whether these are included by default. That is #16807.

**Comment by bradcray** [comment]
I think they should likely be defined in the same module as #16808, whatever it's called.

**Comment by gbtitus** [comment]
I also favor them being in a topic-named module, but I'm across the fence from @bradcray ("Howdy, neighbor!") and think the compilation-time and execution-time things should be in separate modules.

**Comment by e-kayrakli** [comment]
I agree with Brad that compiler/runtime errors/warnings should be grouped in the same module.

@gbtitus -- do you have concerns other than principle that they should be in separate modules? I am thinking of module initialization/resolution order issues, but can't think of one.

**Comment by gbtitus** [comment]
@e-kayrakli and others, no, nothing other than principle, and that rather weak. I'm fine if we end up putting the compile-time and execution-time warnings and errors in the same module.

**Comment by bradcray** [comment]
I just ran into a case that was somewhat interesting to me, and somewhat related to this general topic along with, to an extent, #16808 and #16807:  I was working on some functions that have both param and non-param forms:

```chapel
proc foo(x: int): int { ... }
proc foo(param x: int) param : int { ... }
```

each of which contained an error condition.  Apart from the fact that one error was expressed as a `halt()` and the other a `compilerError()`, the two functions were identical.  At times, we've talked about having a way of writing the function once and having both param and non-param versions of it be created (e.g., #14639) to avoid the code duplication.  For example, imagine:

```chapel
proc foo(param(?isParam) x: int) param(isParam) : int {
  ...do all the stuff...
}
```

where (1) `param([param bool expression]` would be implemented as `param` if the expression were true and `const` if it were not and (2) `param(?b)` could be used to query whether an incoming argument was param or not, similar to `?t` style queries in other contexts.

OK, so if you're with me so far, it seems like it'd be a shame to have the unified function above be a single piece of code that then needs to contain:

```chapel
  if isParam then
    compilerError("Error condition reached");
  else
    halt("Error condition reached");
```

rather than some sort of unified call, for example, imagine:

```chapel
    error("Error condition reached", compiletime=isParam);
```

which will be generated as either a compile-time or execution-time error depending on whether `isParam` is true or false.  In fact, we could _almost_ write such a unified routine today, as:

```chapel
inline proc error(param msg: string, param compiletime: bool = false) {
  if (compiletime) then
    compilerError(msg);
  else
    halt(msg);
}
```

except that (a) the line numbers it reported wouldn't be useful without additional work, and (b) we'd want the `msg` to be able to be non-param for non-compile-time errors.  This could be done through an additional overload:

```chapel
inline proc error(msg: string) {
  halt(msg);
}
```

but nicer might be to express it using the same unified param-query notation:

```chapel
inline proc error(param(?paramMsg) msg: string, param compiletime: bool = false) {
  if (compiletime) then
    if (!paramMsg) then
      compilerError("compile-time errors require `param` error messages");
    else
      compilerError(msg);
  else
    halt(msg);
}
```
